src/hevm/src/EVM.hs

```diff
@@ -2012,6 +2011,79 @@ cheatActions =
+          _ -> vmError (BadCheatCode sig),
+
+      action "file(bytes32,address)" $
+        \sig _ _ input -> case decodeStaticArgs input of
+          [what', addr] -> forceConcrete what' $ \(C _ (utf8Word -> what)) ->
+            case what of
+              "caller" -> assign (state . caller) (SAddr . sFromIntegral . rawVal $ addr)
+              "origin" -> makeUnique addr $ \(C _ (num -> who)) -> assign (tx . origin) who
+              "coinbase" -> makeUnique addr $ \(C _ (num -> who)) -> assign (block . coinbase) who
+              _ -> vmError (BadCheatCode sig)
+          _ -> vmError (BadCheatCode sig),
+
+      action "file(bytes32,uint256)" $
+        \sig _ _ input -> case decodeStaticArgs input of
+          [what', val'] -> forceConcrete2 (what', val') $ \((C _ (utf8Word -> what)), val) ->
+            case what of
+              "gasPrice" -> assign (tx . gasprice) val
+              "txGasLimit" -> assign (tx . txgaslimit) val
+              "blockGasLimit" -> assign (block . gaslimit) val
+              "difficulty" -> assign (block . difficulty) val
+              _ -> vmError (BadCheatCode sig)
+          _ -> vmError (BadCheatCode sig),
+
+      action "look(bytes32,address)" $
+        \sig outOffset _ input ->
+          case decodeStaticArgs input of
+            [what', who'] -> forceConcrete what' $ \(C _ (utf8Word -> what)) ->
+              makeUnique who' $ \(C _ (num -> who))-> do
+                case what of
+                  "nonce" -> do
+                    vm <- get
+                    case Map.lookup who $ view (env . contracts) vm of
+                      Just c' -> do
+                        let out = litWord (_nonce c')
+                        assign (state . returndata . word256At 0) out
+                        assign (state . memory . word256At outOffset) out
+                      _ -> vmError (BadCheatCode sig)
+                  _ -> vmError (BadCheatCode sig)
+            _ -> vmError (BadCheatCode sig),
+
+      action "look(bytes32)" $
+        \sig outOffset _ input -> case decodeStaticArgs input of
+            [what'] -> forceConcrete what' $ \(C _ (utf8Word -> what)) ->
+              case what of
+                "txGasLimit" -> do
+                  vm <- get
+                  let out = litWord $ view (tx . txgaslimit) vm
+                  assign (state . returndata . word256At 0) out
+                  assign (state . memory . word256At outOffset) out
+                _ -> vmError (BadCheatCode sig)
+            _ -> vmError (BadCheatCode sig),
+
+      action "replace(address,bytes)" $
+        \sig _ _ input -> case decodeBuffer [AbiAddressType, AbiBytesDynamicType] input of
+          CAbi vals -> case vals of
+            [AbiAddress usr, AbiBytesDynamic (ConcreteBuffer -> newCode)] -> do
+              vm <- get
+              -- TODO: make error messages nice here
+              when (usr == (view (state . codeContract) vm)) $ vmError (BadCheatCode sig)
```

Comment:
this is kind of an antipattern. We don't want to change anything after we run into a vmError. I think it would be better to change this to a `case` statement or use the continuation passing style applied elsewhere

Reply:
oh, execution will continue after calling `vmError`?

Reply:
the way you currently have it written, yes. `when` is just a shorthand for `if x then blah else return ()` so the state monad continues to the other statements after as well